src/main/java/io/confluent/connect/jdbc/dialect/DatabaseDialect.java

```diff
@@ -0,0 +1,471 @@
+   * Create the definition for the columns described by the database metadata.
+   *
+   * @param connection     the database connection; may not be null
+   * @param catalogPattern the pattern for matching the catalog; may be null
+   * @param schemaPattern  the pattern for matching the schemas; may be null
+   * @param tablePattern   the pattern for matching the tables; may be null
+   * @param columnPattern  the pattern for matching the columns; may be null
+   * @return the column definitions keyed by their {@link ColumnId}; never null
+   * @throws SQLException if there is an error accessing the metadata
+   */
+  Map<ColumnId, ColumnDefinition> describeColumns(
+      Connection connection,
+      String catalogPattern,
+      String schemaPattern,
+      String tablePattern,
+      String columnPattern
+  ) throws SQLException;
+
+  /**
+   * Create the definition for the columns in the result set.
+   *
+   * @param rsMetadata the result set metadata; may not be null
+   * @return the column definitions keyed by their {@link ColumnId} and in the same order as the
+   *     result set; never null
+   * @throws SQLException if there is an error accessing the result set metadata
+   */
+  Map<ColumnId, ColumnDefinition> describeColumns(
+      ResultSetMetaData rsMetadata
+  ) throws SQLException;
+
+  /**
+   * Get the definition of the specified table.
+   *
+   * @param connection the database connection; may not be null
+   * @param tableId    the identifier of the table; may not be null
+   * @return the table definition; null if the table does not exist
+   * @throws SQLException if there is an error accessing the metadata
+   */
+  TableDefinition describeTable(
+      Connection connection,
+      TableId tableId
+  ) throws SQLException;
+
+  /**
+   * Create the definition for the columns in the result set returned when querying the table. This
+   * may not work if the table is empty.
+   *
+   * @param connection the database connection; may not be null
+   * @param tableId    the name of the table; may be null
+   * @return the column definitions keyed by their {@link ColumnId}; never null
+   * @throws SQLException if there is an error accessing the result set metadata
+   */
+  Map<ColumnId, ColumnDefinition> describeColumnsByQuerying(
+      Connection connection,
+      TableId tableId
+  ) throws SQLException;
+
+  /**
+   * Create a criteria generator for queries that look for changed data using timestamp and
+   * incremented columns.
+   *
+   * @param incrementingColumn the identifier of the incremented column; may be null if there is
+   *                           none
+   * @param timestampColumns   the identifiers of the timestamp column; may be null if there is
+   *                           none
+   * @return the {@link TimestampIncrementingCriteria} implementation; never null
+   */
+  TimestampIncrementingCriteria criteriaFor(
+      ColumnId incrementingColumn,
+      List<ColumnId> timestampColumns
+  );
```

Comment:
The timestamp and incrementing criteria is some of the most widely varied we've seen across different DBMSes, so this method allows a dialect to generate this criteria. Other similar methods do the same thing for other kinds of criteria and statements.